visualization/app/codeCharta/state/store/metricData/edgeMetricData/edgeMetricData.reducer.ts

```diff
@@ -0,0 +1,118 @@
+import { EdgeMetricDataAction, EdgeMetricDataActions, setEdgeMetricData } from "./edgeMetricData.actions"
+import { BlacklistItem, BlacklistType, Edge, EdgeMetricCount, EdgeMetricData } from "../../../../codeCharta.model"
+import { getVisibleFileStates } from "../../../../model/files/files.helper"
+import { CodeMapHelper } from "../../../../util/codeMapHelper"
+import { FileState } from "../../../../model/files/files"
+import { EdgeMetricDataService } from "./edgeMetricData.service"
+
+const clone = require("rfdc")()
+
+// Required for performance improvements
+export type NodeEdgeMetricsMap = Map<string, Map<string, EdgeMetricCount>>
+export let nodeEdgeMetricsMap: NodeEdgeMetricsMap
+
+export function edgeMetricData(state: EdgeMetricData[] = setEdgeMetricData().payload, action: EdgeMetricDataAction): EdgeMetricData[] {
+	switch (action.type) {
+		case EdgeMetricDataActions.SET_EDGE_METRIC_DATA:
+			return clone(action.payload)
+		case EdgeMetricDataActions.CALCULATE_NEW_EDGE_METRIC_DATA:
+			return clone(calculateMetrics(action.payload.fileStates, action.payload.blacklist))
+		default:
+			return state
+	}
+}
+
+function calculateMetrics(fileStates: FileState[], blacklist: BlacklistItem[]): EdgeMetricData[] {
+	const hashMap = calculateEdgeMetricData(fileStates, blacklist)
+	const newEdgeMetricData = getMetricDataFromMap(hashMap)
+	addNoneMetric(newEdgeMetricData)
+	sortNodeEdgeMetricsMap()
+	return newEdgeMetricData
+}
+
+function calculateEdgeMetricData(fileStates: FileState[], blacklist: BlacklistItem[]): Map<string, Map<string, EdgeMetricCount>> {
+	nodeEdgeMetricsMap = new Map()
+	const pathsPerFileState = getVisibleFileStates(fileStates).map(fileState => CodeMapHelper.getAllPaths(fileState.file.map))
+	const allFilePaths: string[] = [].concat(...pathsPerFileState)
+	getVisibleFileStates(fileStates).forEach(fileState => {
+		fileState.file.settings.fileSettings.edges.forEach(edge => {
+			if (bothNodesAssociatedAreVisible(edge, allFilePaths, blacklist)) {
+				addEdgeToCalculationMap(edge)
+			}
+		})
+	})
+	return nodeEdgeMetricsMap
```

Comment:
Returning `nodeEdgeMetricsMap` here seems rather confusing since the variables scope is the file.